pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,492 @@
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing failed, requeue", "eip", key)
+	}
+	return true
+}
+
+func (c *Cluster) syncConsistentHash(eipName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing consistentHash", "eip", eipName, "durationTime", time.Since(startTime))
+	}()
+
+	c.consistentHashRWMutex.Lock()
+	defer c.consistentHashRWMutex.Unlock()
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	// updateConsistentHash refresh the consistentHashMap
+	updateConsistentHash := func(eip *v1alpha2.ExternalIPPool) error {
+		nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if err != nil {
+			return err
+		}
+		nodes, err := c.nodeLister.List(nodeSel)
+		if err != nil {
+			return err
+		}
+		c.consistentHashMap[eip.Name] = newNodeConsistentHashMap()
+		aliveNodes := c.aliveNodes()
+		// Node alive and Node labels matches with externalIPPool nodeSelector
+		var aliveAndMatchedNodes []string
+		for _, node := range nodes {
+			nodeName := node.Name
+			if aliveNodes.Has(nodeName) {
+				aliveAndMatchedNodes = append(aliveAndMatchedNodes, nodeName)
+			}
+		}
+		c.consistentHashMap[eip.Name].Add(aliveAndMatchedNodes...)
+		c.notify(eip.Name)
+		return nil
+	}
+
+	if err := updateConsistentHash(eip); err != nil {
+		klog.ErrorS(err, "Sync consistentHash by EIP failed", "EIP", eip.Name)
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	node, event := nodeEvent.Node, nodeEvent.Event
+
+	switch event {
+
+	case memberlist.NodeJoin, memberlist.NodeLeave:
+		// when a Node joins cluster, all matched externalIPPools consistentHash should be updated.
+		// when a Node leave cluster, the Node may failed or deleted,
+		// if Node has been deleted, affected externalIPPool should enqueue, delete node handler has processed
+		// if the Node failed, externalIPPools consistentHash maybe changed, affected externalIPPool should enqueue
+		coreNode, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// means node has been deleted, delete node handler has processed
+				klog.ErrorS(err, "Processing clusterNode events", "eventType", event)
+				return
+			}
+			klog.ErrorS(err, "Processing clusterNode events get node failed", "eventType", event)
+			return
+		}
+		affectedEIPs := c.filterEIPsFromNodeLabels(coreNode)
+		c.enqueueEIPs(affectedEIPs)
+		klog.InfoS("Processed Node event", "eventType", event, "nodeName", node.Name, "eipNum", len(affectedEIPs))
```

Comment:
ditto, let's unify the log message, eipNum => affectedPoolNum

Reply:
done